### PR TITLE
Convos Caterpillar: connective tissue

### PR DESCRIPTION
**Pull request by samouri**
Lets get some 🐛s on convos

currently based from ~~[this branch](https://github.com/Automattic/wp-calypso/pull/17589)~~ master.

To Do:
- [x] land blocker PRs as specified in https://github.com/Automattic/wp-calypso/pull/17517
- [x] fix how caterpillars (don't) handle nested children.  i.e. expanding multiple layers up to N
- [ ] ~~animation for moving the faces? cc @fraying~~ deferring to future PR
- [x] address issues with devdocs

### CODE REVIEW

**Comment by fraying** [comment]
Nice! I'd like to come back to animations once we've got the functionality working.

**Comment by samouri** [comment]
Note: I plan on deferring what to do about posts with 50+ comments to later.  Right now caterpillar is only capable of loading the first 50

**Comment by jancavan** [comment]
> I'd like to come back to animations once we've got the functionality working.

I'd like to help with this! I think some animation is necessary because I lose track of where I am after clicking the caterpillar.

File: client/blocks/comments/post-comment.jsx [hunk 1]
```diff
@@ -45,6 +46,8 @@ class PostComment extends Component {
 		maxDepth: PropTypes.number,
 		showNestingReplyArrow: PropTypes.bool,
 		displayType: PropTypes.oneOf( values( POST_COMMENT_DISPLAY_TYPES ) ),
+		toShow: PropTypes.object, // if provided, hide comments not specified in this object
```
Comment (bluefuton) [inline client/blocks/comments/post-comment.jsx]: What's the structure of this object?

File: client/blocks/conversation-caterpillar/style.scss [hunk 2]
```diff
@@ -1,6 +1,14 @@
 .conversation-caterpillar.card {
 	cursor: pointer;
 	display: flex;
+	padding: 0px;
```
Comment (bluefuton) [inline client/blocks/conversation-caterpillar/style.scss]: Zero doesn't need units :)

**Comment by bluefuton** [comment]
Looks good! 

My feedback is all about UX and that's something we can refine later:
- Does the caterpillar do enough to invite interaction?
- I think we need to add clearer information about what's happening on each caterpillar click (clicking will expand x comments).

File: client/blocks/comments/post-comment.jsx (hunk 1 above)
Comment (bluefuton) [reply client/blocks/comments/post-comment.jsx]: I would be explicit and call it `commentsToShow` I think.

File: client/blocks/comments/post-comment.jsx [hunk 3]
```diff
@@ -196,11 +224,34 @@ class PostComment extends Component {
...
-			? POST_COMMENT_DISPLAY_TYPES.full
-			: this.props.displayType;
+		const isPingbackOrTrackback = comment.type === 'trackback' || comment.type === 'pingback';
+
+		if ( ! comment || ( this.props.hidePingbacksAndTrackbacks && isPingbackOrTrackback ) ) {
+			return null;
+		} else if ( toShow && ! toShow[ commentId ] ) {
+			// this comment should be hidden so just render children
+			return (
+				<div>
```
Comment (bluefuton) [inline client/blocks/comments/post-comment.jsx]: Is it possible that the replies list will be empty here? @jancavan reported empty divs being rendered:

screenshot_2017-08-30_14 54 18_360

Comment (samouri) [reply client/blocks/comments/post-comment.jsx]: hmmm.....can do a check for if any children should be rendered and skip the divs that don't apply. I'm worried its more work than rendering empty divs.

I'll get rid of the divs for now but if perf problems arise we'll want to store smarter data structures

File: client/blocks/comments/post-comment.jsx (hunk 1 above)
Comment (samouri) [reply client/blocks/comments/post-comment.jsx]: Good question! I'll throw in some comments

```
{
  [ commentId ]: displayType
}
```

**Comment by samouri** [comment]
UX nit I noticed: if a caterpillar is hiding only a single comment, it takes up as much space as the comment. Lets skip the ceremony and expand the comment

**Comment by samouri** [comment]
> Should expansion state persist? I expected the caterpillars to be back in place when I revisited /conversations, mainly for testing purposes I admit ;)

Thats a good question.  it'll be relatively easy to implement something that unexpands all comments when you load a stream just like we did for [video cards](https://github.com/Automattic/wp-calypso/blob/master/client/state/ui/reader/card-expansions/reducer.js#L24)

File: client/blocks/comments/post-comment.jsx (hunk 3 above)
Comment (samouri) [reply client/blocks/comments/post-comment.jsx]: fixed with https://github.com/Automattic/wp-calypso/pull/17608/commits/a5326ec20a2510ebba93a394b5eb5c144aa893f4

File: client/blocks/conversation-caterpillar/style.scss (hunk 2 above)
Comment (samouri) [reply client/blocks/conversation-caterpillar/style.scss]: done

File: client/blocks/comments/post-comment.jsx (hunk 1 above)
Comment (samouri) [reply client/blocks/comments/post-comment.jsx]: renamed and comments added in:  https://github.com/Automattic/wp-calypso/pull/17608/commits/5dec64945d07e90f37d373fb2742f5eda70f8bda

**Comment by samouri** [comment]
Thanks for the helpful and insightful feedback @bluefuton!

While writing out the explanation for `commentsToShow` i realized we can deprecate the short-lived `displayType` prop considering it can be fully gleaned from the `commentsToShow` object which should probably also be renamed to something more precise (it isn't just which ones to show/hide, but also the displayType).

I think that should be for future PRs though since this is pretty big as-is and I think we are going to have tweaks to the caterpillar implementation for weeks to come

**Comment by bluefuton** [comment]
Thanks for the changes @samouri!

I've noticed one last regression - some comments getting cropped since we added caterpillars.

On wpcalypso:

screen shot 2017-08-31 at 21 37 18

On this branch:

screen shot 2017-08-31 at 21 37 28

**Comment by samouri** [comment]
@bluefuton: thats by design! Notice the "Read More" link?  Theres a fade/crop that happens in specific circumstances.  We plan on iterating on how exactly the fade happens in the near future.

**Comment by bluefuton** [comment]
@samouri I noticed the 'read more' but the crop looked a bit severe! If we're going to iterate on that, fine to disregard my note for now 👍

**Comment by bluefuton** [comment]
Do we have a todo issue capturing what's left to add to caterpillars?

**Comment by samouri** [comment]
@bluefuton: i'll make some issues now!

**Comment by fraying** [comment]
I don't understand this comment: 

> if a caterpillar is hiding only a single comment, it takes up as much space as the comment. Lets skip the ceremony and expand the comment

A comment is twice the height of a caterpillar even when the comment is only one line, and most comments are more than one line. (Don't forget the comment footer links.) So, it's not true that it takes up as much space. Anyway, the point of the caterpillars is to not show things if they're old or you've seen them before, regardless of the space savings. So, no, please do not expand comments if there's just one in the caterpillar.

**Comment by fraying** [comment]
@bluefuton 

>Should expansion state persist? I expected the caterpillars to be back in place when I revisited /conversations, mainly for testing purposes I admit ;)

I get why we might want that during testing, but we definitely don't want to persist the state of expanded comments. The expansion is to view the comment. Once that happens, we can assume the comment has been viewed, which means on the next load we want to hide that comment specifically. The goal is to keep the page compact and only expand the new things, so the user's eye is drawn to what's new, not what's already been seen.

**Comment by fraying** [comment]
Note that there is a "view all" link in the designs we haven't implemented yet, which would be a quick way to load in ALL comments for a particular post.

**Comment by samouri** [comment]
@fraying: 

> I don't understand this comment:

Sorry for jumping to a solution, I should have started with what I think could be a problem.
Heres a gif of me interacting with a caterpillar that sparked the thought

single-comment

**Comment by fraying** [comment]
@samouri I getcha. And, yeah, in that example, there's not much to be gained by hiding the comment in a caterpillar. But what happens if the comment is longer (as most are)? Space is at a premium in Conversations, so I still think we should hide it. Remember, ideally, we're only hiding it because the user has already read it. So it's there to show that someone commented there, and the user can expand it if they want to, but it makes sense to minimize it. (At least, I think so.)

**Comment by samouri** [comment]
> Do we have a todo issue capturing what's left to add to caterpillars?

While we don't have an explicit todo, we _do_ have a [milestone](https://github.com/Automattic/wp-calypso/milestone/202) that I believe has all the main caterbits left

**Pull request merged**